crates/sol-type-parser/src/parameter.rs

```diff
@@ -0,0 +1,202 @@
+use crate::{opt_ws_ident, spanned, tuple_parser, Error, Result, TypeSpecifier};
+use alloc::vec::Vec;
+use winnow::{trace::trace, PResult, Parser};
+
+/// Represents a function parameter.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct ParameterSpecifier<'a> {
+    /// The full span of the specifier.
+    pub span: &'a str,
+    /// The type of the parameter.
+    pub ty: TypeSpecifier<'a>,
+    /// Whether the parameter indexed.
+    pub indexed: bool,
+    /// The name of the parameter.
+    pub name: Option<&'a str>,
+}
+
+impl<'a> TryFrom<&'a str> for ParameterSpecifier<'a> {
+    type Error = Error;
+
+    #[inline]
+    fn try_from(value: &'a str) -> Result<Self> {
+        Self::parse(value)
+    }
+}
+
+impl<'a> ParameterSpecifier<'a> {
+    /// Parse a parameter from a string.
+    #[inline]
+    pub fn parse(input: &'a str) -> Result<Self> {
+        Self::parser.parse(input).map_err(Error::parser)
+    }
+
+    pub(crate) fn parser(input: &mut &'a str) -> PResult<Self> {
+        trace(
+            "ParameterSpecifier",
+            spanned(|input: &mut &'a str| {
+                let ty = TypeSpecifier::parser(input)?;
+                let mut indexed = false;
+                let mut name = opt_ws_ident(input)?;
+                // TODO: ?
+                // if let Some("storage" | "memory" | "calldata") = name {
+                //     name = opt_ws_ident(input)?;
+                // }
```

Comment:
Should we add this?

Reply:
I wonder if e.g. you were building Stylus maybe you'd need this to differentiate between when to copy calldata vs not, but I think we can leave as a todo? 

alternatively it might be a simple matter of having an enum called `pointer: Storage | Memory | Calldata` as part of the struct like you did with the indexed events?

Reply:
Added in 9be37b13050d1a3d246c873bb3f7a87354dd1f5e